### PR TITLE
fix(typescript): fixes #1652 compile when source exists anywhere in the working directory

### LINKED ISSUE #1652: Typescript plugin fails to compile the entry point when there is at least one invalid TypeScript file somewhere is the working directory
- Rollup Plugin Name: @rollup/plugin-typescript
- Rollup Plugin Version: 11.1.5
- Rollup Version: 4.9.0
- Operating System (or Browser): Ubuntu
- Node Version: 16.20
- Link to reproduction: https://github.com/ericmorand/rollup-plugin-typescript-issue-2

### Expected Behavior

In the reproduction repository, we find 3 TS files:

* `src/index.ts`, the entry point
* `src/valid.ts`, imported by the entry point
* `src/invalid.ts`, containing non valid TypeScript syntax and imported by neither `src/index.ts` or `src/valid.ts`

When executing rollup on `src/index.ts`, the build should be successful since the `src/invalid.ts` file is not imported.

### Actual Behavior

The build is unsuccessful, and the error suggests that `src/invalid.ts` is syntax-checked by the plugin even though it is not part of the dependency graph of the entry point:

```shell
$ npx rollup -i src/index.ts -p @rollup/plugin-typescript
npm WARN cli npm v10.2.4 does not support Node.js v16.20.2. This version of npm supports the following node versions: `^18.17.0 || >=20.5.0`. You can find the latest version at https://nodejs.org/.

src/index.ts → stdout...
var foo = 5;

console.log(foo);
(!) Plugin typescript: @rollup/plugin-typescript TS2304: Cannot find name 'foo'.
src/invalid.ts: (1:1)

1 foo
  ~~~
```

### Additional Information

Note that the behaviour of `tsc` is correct in that regard:

```shell
$ tsc src/index.ts
```

No error is emitted and `src/index.ts` and `src/valid.ts` are successfully compiled.

So the error is coming from the plugin, not from the TypeScript compiler.

### PR DESCRIPTION
**Pull request by ericmorand**
## Rollup Plugin Name: `typescript`

This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

- resolves #1652

### Description

This PR solves the issue described in #1652. It focuses on not creating a breaking changes by ensuring that no existing test was modified - except one that I'd like to discuss with the maintainers.

Basically, what this PR does is to remove the initial emit of **all** the files included in th working directory and instead emit files that are part of the dependency graph of the entry point.

### CODE REVIEW

**Review by shellscape — APPROVED** [review]
I think the changes here look good. Thanks for putting this together. I do think this needs another person to put eyes on it before merging. @NotWoods please take a look if you have the chance.

**Comment by shellscape** [comment]
Apologies for the delay. We always have a shortage of maintainers for this repo, and it doesn't look like anyone else has stepped in to review this. Best I can do is give them another ping.
